### PR TITLE
Fix 317

### LINKED ISSUE #317: Allow users to download the Land cover TIFF file after clipping with the catchment boundaries
Users would like to be able to get the clipped land-use raster for their catchment along with the % of each land-use category. Adding this as an option out of the NALCMS zonal stats WPS would be perfect.

### PR DESCRIPTION
**Pull request by Zeitsperre**
## Overview

This PR fixes #317 

Changes:

* Refactored out the core GIS functions for several GIS processes to the `ravenpy` library (now all in version 0.3.0).
* Added `wps_nalcms_zonal_stats_raster.py`. This process is `nalcms-zonal-stats` with raster output from the core functionality of `gis.zonalstats_raster_file()`.
* Performed some quality of life changes (`black`/`isort`/f-strings).

## Additional Information

Links to other issues or sources.

### CODE REVIEW

**Review by huard — COMMENTED** [review]
Will take a deeper look later, but note that for Inputs and Outputs that are reused across processes, you can define them in wpsio.

File: raven/utils.py
Comment (huard) [inline raven/utils.py:817]: `file` is a python type. Use `fn`

File: raven/utils.py
Comment (huard) [inline raven/utils.py:800]: Do you really need `identifier` or could it be a random name ?

Comment (huard) [reply raven/utils.py:800]: Remove reference to self.

File: raven/processes/wps_nalcms_zonal_stats_raster.py
Comment (huard) [inline raven/processes/wps_nalcms_zonal_stats_raster.py:123]: Move identical inputs used across multiple processes to wpsio.py.

**Review by huard — APPROVED** [review]
Could you add a very short notebook using the new process and plotting it ?

File: raven/processes/wps_generic_raster_subset.py
Comment (Zeitsperre) [reply raven/processes/wps_generic_raster_subset.py:198]: Yes. Was just holding onto it in case.

File: raven/utils.py
Comment (Zeitsperre) [reply raven/utils.py:817]: True that!

File: raven/utils.py
Comment (Zeitsperre) [reply raven/utils.py:800]: It could be a random name, certainly. Would it makes sense to just have a function make a random string and remove this field ?

Comment (huard) [reply raven/utils.py:800]: I think so.

File: raven/processes/wps_nalcms_zonal_stats_raster.py
Comment (Zeitsperre) [reply raven/processes/wps_nalcms_zonal_stats_raster.py:123]: Will do.

Comment (Zeitsperre) [reply raven/processes/wps_nalcms_zonal_stats_raster.py]: Done.

**Comment by Zeitsperre** [comment]
@huard  I'm tempted to run `black` on `wpsio.py`, but I want to give you a chance to look it over before the changes are too enormous. All tests are passing on my end.

If everything is looking good from your perspective, I'll add a notebook (or modify an existing one) to make use of the new process, and we'll go from there.

What's the plan on porting updates to RavenPy by the way? I can see there being a lot of use in refactoring `utils.py`, `gis.py`, and `wpsio.py`. I'm wondering how we plan on doing this (all at once? Gradually?).

**Comment by huard** [comment]
I suggest we run black over the entire code base in another PR after the switch to RavenPy in a dedicated PR. 

Please go ahead with the notebook. 

I think we'll port changes from the recent PRs to RavenPy, then modify Raven to depend on RavenPy. 
I suggest we do any refactoring in small bits, easier to review and debug.

**Comment by Zeitsperre** [comment]
@huard 

Running into a snag here: I realized that some processes returning zip files is really clunky (also, Jupyter notebooks is very careful about saving bytes/potentially executable code), so I'm looking into the METALINKv4 methods, but I can't find any means of reading the data it links to.

image

It may be the case that the GIS processes that can (potentially) return more than one GeoTIFF (there are at least 2-3 I can think of) may need to be rewritten using the PyWPS-supported transport methods (and not a zip file). Do you have any experience with creating/working with these outputs?

File: raven/processes/wps_nalcms_zonal_stats_raster.py
Comment (huard) [inline raven/processes/wps_nalcms_zonal_stats_raster.py:25]: That's already defined elsewhere, no ? Why not import it ?

Comment (Zeitsperre) [reply raven/processes/wps_nalcms_zonal_stats_raster.py:25]: Good point. Will do!

**Comment by huard** [comment]
Is pymetalink installed ? 
The birdy client has an importer for MetaLink document that fetches its content. However, there is no importer for TIFF files. Adding one would require adding a class similar to this: https://github.com/bird-house/birdy/blob/ed4998154fd723a079c1b5bb773c860cd5affcdb/birdy/client/converters.py#L227

**Comment by huard** [comment]
Note that not having a geotiff importer means that the client will return the URL to the geotiff. If there is a well maintained option to read geotiff, we could add an importer to birdy.

**Comment by Zeitsperre** [comment]
@huard I'm coming back to this PR today. Changes in https://github.com/CSHS-CWRA/RavenPy/pull/39 should not break anything wrt `utils.py` and `gis.py`, but once that PR is ready, I'd like to start re-arranging things in RavenPy which will require refactoring on RAVEN.

**Comment by Zeitsperre** [comment]
@richardarsenault @huard 

This PR is nearly done. It currently requires the new mimetype definitions coming in the new PyWPS (geoTIFF and GeoJSON). It also makes use of newer `Proj` binaries, so CRS transformations now raise fewer warnings and should be much less prone to buggy behaviour. For backwards compatibility, there is now a `NALCMSZonalStatisticsRasterProcess()` that does the exact same thing as `NALCMSZonalStatisticsProcss()` except it also returns raster files (one for each feature, accesible via `resp["raster"]`) in a [PyMetalink](https://github.com/metalink-dev/pymetalink) XML reference. 

From my understanding, this type of object uses some neat magic with converters in Jupyter Notebooks to access the linked files(s). Outside this, the method of accessing them in python scripts is as follows:
```
from metalink import download as md
from common import get_output
out = get_output(resp.xml)
d = md.get(out['output'], path='/tmp')
```
`PyMetalink` is not installed by default in raven (but I've started a conda recipe for it with @cehbrecht https://github.com/conda-forge/staged-recipes/pull/13897).

Once the new PyWPS is integrated, ensuring the notebooks are passing will be a next step. @huard for review, If you could please try the `NALCMSZonalStatisticsRasterProcess()` yourself, that would be helpful. Thanks again!

**Comment by huard** [comment]
@Zeitsperre Does it work on your end ? I got failures here that I tracked to a missing trailing / in ravenpy GEO_URL.

**Comment by Zeitsperre** [comment]
All tests (except for `TestNALCMSZonalStatsWithRasterProcess`, due to a metalink file-access error) appear to be passing for me. Perhaps we can discuss this soon.

**Comment by Zeitsperre** [comment]
This PR is now dependent on changes in RavenPy (https://github.com/CSHS-CWRA/RavenPy/pull/68), Any refactoring there will demand changes here.

**Comment by Zeitsperre** [comment]
@huard I've updated the PyWPS to 4.4.0, yest I still keep running into the same issue for metalink file access:
```
...
  File "/home/tjs/miniconda3/envs/raven/lib/python3.8/urllib/request.py", line 1425, in unknown_open
    raise URLError('unknown url type: %s' % type)
urllib.error.URLError: <urlopen error unknown url type: ?xml version="1.0" encoding="utf-8"?>
<metalink xmlns="urn>
```
The XML is structured fine and urllib3 should not have issues downloading data from `file:///` addresses. Here's whole XML string:
```
<ns0:metalink xmlns:ns0="urn:ietf:params:xml:ns:metalink">
    <ns0:published>2021-03-11T15:50:54Z</ns0:published>
    <ns0:generator>PyWPS/4.4.0</ns0:generator>
    <ns0:file name="subset_1.tiff">
        <ns0:identity>subset_1.tiff</ns0:identity>
        <ns0:size>895973</ns0:size>
        <ns0:metaurl mediatype="image/tiff; subtype=geotiff">file:///tmp/78ac79c8-82ab-11eb-92b7-28d24417f566/subset_1.tiff</ns0:metaurl>
        <ns0:publisher name="None" url="http://localhost/wps" />
    </ns0:file>
</ns0:metalink>
```
This looks perfectly fine to me. Any ideas?

**Comment by huard** [comment]
All tests in test_wps_nalcms_zonal_stats.py pass on my end.

**Comment by huard** [comment]
Wait, updated pymetalink and now I get the same error.

**Comment by richardarsenault** [comment]
I've found 2 possibilities, if that can help any:

A few solutions people have found: 
https://stackoverflow.com/questions/28376506/urllib-https-request-urlopen-error-unknown-url-type-https

Another related to the way the string is written:
https://stackoverflow.com/questions/27115803/urllib-error-urlerror-urlopen-error-unknown-url-type-https

Hope that may help, maybe.

**Comment by huard** [comment]
Solved. `md.get`  is made to first retrieve the metalink file, then download its content. You were calling get on the metalink file content directly.

**Comment by Zeitsperre** [comment]
@huard Thanks for the help! With this change and all the PRs for support behaviour, I can finally finish the work I started in... December 2020. very exciting!

**Comment by Zeitsperre** [comment]
Also worth noting: This PR should now close #364 with https://github.com/Ouranosinc/raven/pull/335/commits/326cd3437c9c34f75d6796d7d900a471ef1f5541.

**Comment by Zeitsperre** [comment]
@huard This PR is in working order. If you'd like to give it a final look-over, feel free. Additional changes will be more on par with tweaks.

One thing worth returning to eventually is the desired converter for GeoTIFF files (Birdy). As it stands, asking for an object will return the grid as a `numpy.array`. Masking the array requires much more information about the data contents, so this should be performed manually.

What I think would be a neat is to convert the data with rioxarray (`xarray.open_rasterio` + `rio` module). This would give a lot of control in notebooks. To be discussed eventually.

**Comment by huard** [comment]
Ok. Please go ahead and merge, writing a commit message that synthesizes the PR.

**Comment by huard** [comment]
The PRs in emu and birdy are all green and ready to be merged.

**Comment by Zeitsperre** [comment]
The last thing I want to fix are the two XFAILED tests that calculate area (`test_wps_generic_shape_properties::TestGenericShapePropertiesProcess::test_multifeature_*`). I've verified with QGIS that the area values should be what is expected but yet they seem off by a fairly large amount (~100 sq. m.).

Once I get to the bottom of this, I'll merge. Thanks again!

**Comment by Zeitsperre** [comment]
I think I've finally figured out the reason behind the value inconsistencies for the affected tests (test_wps_generic_shape_properties::TestGenericShapePropertiesProcess::test_multifeature_*). Summarizing my findings here.

The transformation from EPSG:4326 to EPSG:6622 involves a geoid transformation (WGS84 -> GRS80), and there are multiple methods for performing it with different levels of accuracies. I've determined that the transformation used is this one:
```
+proj=pipeline +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=push +v_3 +step +proj=cart +ellps=WGS84 +step +inv +proj=helmert +x=-0.991 +y=1.9072 +z=0.5129 +rx=-0.0257899075194932 +ry=-0.0096500989602704 +rz=-0.0116599432323421 +s=0 +convention=coordinate_frame +step +inv +proj=cart +ellps=GRS80 +step +proj=pop +v_3 +step +proj=lcc +lat_0=44 +lon_0=-68.5 +lat_1=60 +lat_2=46 +x_0=0 +y_0=0 +ellps=GRS80
```
I've adjusted to expected values to reflect this and re-run the disabled tests, updating the values for their new geometries. If we were to return to this, changing the choice of transformation might be worthwhile, but all is working now.

**Comment by Zeitsperre** [comment]
@richardarsenault I believe this finally addresses #317.

We now have `wps.nalcms_zonal_stats_raster` as a function that replicates `wps.nalcms_zonal_stats` and also returns the raster file. Eventually this would replace that function, but I'm leaving it for the sake of backwards compatibility.

The notebook `Extract_geographical_watershed_properties.ipynb` shows how to make use of the new process for extracting the data from it. The last thing left to do is fix a bug with the converters that makes it only return the geotiff as a numpy array (which is nice, but an `xarray`-like converter would be better. This work for you?

**Comment by richardarsenault** [comment]
@Zeitsperre This makes total sense, thanks a lot! Looks good! Didn't think this would require so many changes!

**Comment by Zeitsperre** [comment]
@huard For versioning, this raises the version of `RavenPy` needed to 0.3.0 and refactors out quite a bit of the utility code. Should we be bumping `raven-wps` to a new minor?

**Comment by richardarsenault** [comment]
I have a demo on monday where I will want to use JupyterLab notebooks, so please don't break it yet :)

**Comment by Zeitsperre** [comment]
`RavenPy` is no longer compatible with `raven-wps`, so this PR is needed for any new builds. Should I wait past Monday for this PR?

**Comment by richardarsenault** [comment]
I mean, maybe we can merge this without a hitch, I just have no idea if doing so will make the notebooks on JupyterLab obsolete or whatnot. Maybe there is an extra step to update those notebooks on the server that we could just... not do until post-monday?

**Comment by Zeitsperre** [comment]
If everything is working for you right now. I don't want to risk breaking anything in production. I know that tagged releases will trigger rebuilds, so merging this is probably fine so long as we don't tag a version. 

Am I right about this @tlvu ? If so, merging this to master should be OK.

**Comment by tlvu** [comment]
> If everything is working for you right now. I don't want to risk breaking anything in production. I know that tagged releases will trigger rebuilds, so merging this is probably fine so long as we don't tag a version.
> 
> Am I right about this @tlvu ? If so, merging this to master should be OK.

In prod, we use 2 things from this repo:

1. Raven tutorial notebooks from `master` branch so merging this PR will deploy new version of notebooks automatically.
1. Raven docker image from git tag only so merging this PR without making a release won't deploy anything.

I have a feeling @richardarsenault want to keep the notebooks unchanged until after Monday so this PR should not be merged until after Monday.

But if Richard just do not want to deploy a new Raven-WPS version then merging this PR won't automatically deploy.

**Comment by richardarsenault** [comment]
OK then I suggest we wait until monday night/tuesday AM to merge, just in case. I will be using the notebooks on the server so if they update and there are bugs of any kind, it will be problematic. Might as well wait if you are all OK with this.

**Comment by Zeitsperre** [comment]
Thanks @tlvu for the info!

I'll keep this open for now and move on to the birdy/rioxarray converter. It would be a huge gain to have that (and it's probably easy to get it working).

File: raven/processes/wps_generic_raster_subset.py
Comment (cjauvin) [inline raven/processes/wps_generic_raster_subset.py:130]: An f-string inside an f-string is hard to read I find.. and in this case I think it should be:
```python
f"Failed to perform raster subset using {shape_url}{f' and {raster_url} ' if raster_url else ''}: {e}"
```

File: raven/processes/wps_generic_zonal_stats.py
Comment (cjauvin) [inline raven/processes/wps_generic_zonal_stats.py:112]: Same comment as above here

File: raven/processes/wps_nalcms_zonal_stats_raster.py
Comment (cjauvin) [inline raven/processes/wps_nalcms_zonal_stats_raster.py:86]: Maybe put the comment above the line, to prevent Black from doing this ugly thing.

File: raven/processes/wps_nalcms_zonal_stats_raster.py
Comment (cjauvin) [inline raven/processes/wps_nalcms_zonal_stats_raster.py:135]: Maybe replace this with a one-liner?

File: raven/processes/wps_nalcms_zonal_stats_raster.py
Comment (cjauvin) [inline raven/processes/wps_nalcms_zonal_stats_raster.py:201]: Same comment as earlier

File: raven/processes/wps_raven.py
Comment (cjauvin) [inline raven/processes/wps_raven.py:26]: if you insist.. :stuck_out_tongue:

**Review by cjauvin — APPROVED** [review]

**Pull request merged**
